ir/numbers.go

```diff
@@ -54,3 +57,90 @@ func IsEqualNumber(x, y Number) bool {
 	}
 	panic("bug: unknown number type")
 }
+
+func unmarshalInt(p []byte, n *Int) error {
+	if string(p) == "null" {
+		return nil
+	}
+	var z big.Int
+	_, ok := z.SetString(strings.Replace(string(p), "\"", "", -1), 10)
+	if !ok {
+		return fmt.Errorf("not a valid big integer int: %s", p)
+	}
+	n.Int = &z
+	return nil
+}
+
+func unmarshalFloat(p []byte, n *Float) error {
+	if string(p) == "null" {
+		return nil
+	}
+	// Using precision 53 to accommodate for big.NewFloat which uses 53 preciosion by default.
+	// If we change this marshalling behavior to use MarshalText we need to change to 64 precision.
+	// Check: https://github.com/golang/go/issues/45309
+	z, _, err := big.ParseFloat(strings.Replace(string(p), "\"", "", -1), 10, 53, big.ToNearestEven)
+	if err != nil {
+		return err
+	}
+	n.Float = z
+	return nil
+}
+func (n *Float) UnmarshalJSON(b []byte) error {
+	var objMap map[string]*json.RawMessage
+	err := json.Unmarshal(b, &objMap)
+	if err != nil {
+		err = unmarshalFloat(b, n)
+
+		if err != nil {
+			return err
+		}
+	} else {
+
+		if _, ok := objMap["type"]; ok {
+			err = unmarshalFloat(*objMap["value"], n)
+			if err != nil {
+				return err
+			}
+		}
+
+	}
+	return nil
+}
+
+func (n *Int) UnmarshalJSON(b []byte) error {
+	var objMap map[string]*json.RawMessage
+	err := json.Unmarshal(b, &objMap)
+	if err != nil {
+		err = unmarshalInt(b, n)
+
+		if err != nil {
+			return err
+		}
+	} else {
+
+		if _, ok := objMap["type"]; ok {
+			err = unmarshalInt(*objMap["value"], n)
+			if err != nil {
+				return err
+			}
+		}
+
+	}
+	return nil
+}
+
+func (n Int) MarshalJSON() (b []byte, e error) {
+	c := struct {
+		Type  MarshalType `json:"type"`
+		Value string      `json:"value"`
+	}{Type: IntType, Value: n.String()}
```

Comment:
I thought we said the Value field here will equal the output of big.Int.MarshalText().